pkg/topology/scheduler/balancer/balancer.go

```diff
@@ -0,0 +1,420 @@
+		if minUtilization > utilization {
+			minUtilization = utilization
+			storagePool = sp.Name
+		}
+	}
+
+	if storagePool == "" {
+		return "", fmt.Errorf("Something went wrong couldn't calculate utilization for StoragePool")
+	}
+	return storagePool, nil
+}
+
+func pickStoragePoolFromNodes(ctx context.Context, nClient NodeLinstorClient, nodes []string) (*BalanceDecision, error) {
+
+	util, err := getNodesUtil(ctx, nClient, nodes)
+	if err != nil {
+		return nil, err
+	}
+
+	node, err := getLessUsedNode(util)
+	if err != nil {
+		return nil, err
+	}
+
+	prefNic, err := getLessUsedNic(node)
+	if err != nil {
+		return nil, err
+	}
+
+	storagePool, err := getLessUsedStoragePool(prefNic)
+	if err != nil {
+		return nil, err
+	}
+
+	return &BalanceDecision{
+		NodeName:        node.Name,
+		StoragePoolName: storagePool,
+	}, nil
+}
+
+// pick from Storage Nodes in the same Rack
+func pickStoragePoolTopo(ctx context.Context, selectedNode string, nClient NodeLinstorClient, clientset kubernetes.Interface) (sp *BalanceDecision, err error) {
+	rack, err := retrieveRackId(clientset, selectedNode)
+	if err != nil {
+		return nil, err
+	}
+
+	nodes, err := getStorageNodesInRack(rack, clientset)
+	if err != nil {
+		return nil, err
+	}
+
+	return pickStoragePoolFromNodes(ctx, nClient, nodes)
+
+}
+
+// pick from All Storage Nodes
+func pickStoragePool(ctx context.Context, nClient NodeLinstorClient, clientset kubernetes.Interface) (sp *BalanceDecision, err error) {
+	nodes, err := getStorageNodes(clientset)
+	if err != nil {
+		return nil, err
+	}
+
+	nodeList := []string{}
+	for _, node := range nodes.Items {
+		nodeList = append(nodeList, node.Name)
+	}
+
+	return pickStoragePoolFromNodes(ctx, nClient, nodeList)
+
+}
+
+type BalanceScheduler struct {
+	log *logrus.Entry
+	*lc.HighLevelClient
+	clientset kubernetes.Interface
+}
+
+func NewScheduler(c *lc.HighLevelClient, log *logrus.Entry) (b BalanceScheduler, err error) {
+	clientset, err := k8sClient()
+	if err != nil {
+		return b, err
+	}
+
+	return BalanceScheduler{
+		log:             log,
+		HighLevelClient: c,
+		clientset:       clientset,
+	}, nil
+}
+
+func (b BalanceScheduler) deploy(ctx context.Context, vol *volume.Info, params volume.Parameters, node string, storagePool string) error {
+	return b.Resources.Create(ctx, volToDiskfullResourceCreate(vol, params, node, storagePool))
+}
+
+func (b BalanceScheduler) Create(ctx context.Context, vol *volume.Info, req *csi.CreateVolumeRequest) error {
+	topos := req.GetAccessibilityRequirements()
+	if topos == nil {
+		return fmt.Errorf("no volume topologies, unable to schedule volume %s", vol.ID)
+	}
+
+	params, err := volume.NewParameters(vol.Parameters)
+	if err != nil {
+		return fmt.Errorf("unable to create volume due to bad parameters %+v: %v", vol.Parameters, err)
+	}
+
+	if !params.AllowRemoteVolumeAccess {
+		return fmt.Errorf("placementPolicyBalance cannot work on on local storage")
+	}
+
```

Comment:
It might be good to add a guard clause if the `params.StoragePool != ""` since this scheduler picks a storage pool automatically and if the user has specified a storage pool it might indicate they expect different behavior.

Reply:
:+1: